modules/mesh-task/variables.tf

```diff
@@ -106,41 +106,101 @@ variable "envoy_image" {
   type        = any
   default     = null
 }
 
 variable "container_definitions" {
-  description = "Application container definitions (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definitions)."
+  description = "Application [container definitions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definitions)."
   # This is `any` on purpose. Using `list(any)` is too restrictive. It requires maps in the list to have the same key set, and same value types.
   type = any
 }
 
 variable "upstreams" {
   description = <<-EOT
-  Upstream services this service will call. In the form [{destination_name = $name, local_bind_port = $port}] where
-  destination_name is the name of the upstream service and local_bind_port is the local port that this application should
-  use when calling the upstream service.
+  Upstream services that this service will call. This follows the schema of the `proxy.upstreams` field of the
+  [consul-ecs config file](https://github.com/hashicorp/consul-ecs/blob/main/config/schema.json).
   EOT
 
-  type = list(
-    object({
-      destination_name = string
-      local_bind_port  = number
-    })
-  )
+  type    = any
   default = []
+
+  validation {
+    error_message = "Upstream fields 'destinationName' and 'localBindPort' are required."
+    condition = alltrue(flatten([
+      for upstream in var.upstreams : [
+        can(lookup(upstream, "destinationName")),
+        can(lookup(upstream, "localBindPort")),
+      ]
+    ]))
+  }
+
+  validation {
+    error_message = "Upstream fields must be one of 'destinationType', 'destinationNamespace', 'destinationName', 'datacenter', 'localBindAddress', 'localBindPort', 'config', or 'meshGateway'."
+    condition = alltrue(flatten([
+      for upstream in var.upstreams : [
+        for key in keys(upstream) : contains(
+          [
+            "destinationType",
+            "destinationNamespace",
+            "destinationName",
+            "datacenter",
+            "localBindAddress",
+            "localBindPort",
+            "config",
+            "meshGateway",
+          ],
+          key
+        )
+      ]
+    ]))
+  }
 }
 
 variable "checks" {
-  description = "A list of maps defining Consul checks for this service (https://www.consul.io/api-docs/agent/check#register-check)."
-  type        = list(any)
+  description = "A list of maps defining Consul checks for this service. See the Consul [Register Check documentation](https://www.consul.io/api-docs/agent/check#register-check)."
+  type        = any
   default     = []
+
+  validation {
+    error_message = "Check fields must be one of 'checkId', 'name', 'scriptArgs', 'items', 'interval', 'timeout', 'ttl', 'http', 'header', 'method', 'body', 'tcp', 'status', 'notes', 'tlsServerName', 'tlsSkipVerify', 'grpc', 'grpcUseTls', 'aliasNode', 'aliasService', 'successBeforePassing', or 'failuresBeforeCritical'."
```

Comment:
Nice work. Coming up with the exact required arguments must have been a ton of work